tds/modules/dataset/controller.py

```diff
@@ -0,0 +1,176 @@
+"""
+TDS Dataset
+"""
+import os.path
+from logging import Logger
+from typing import List
+
+import boto3
+from elasticsearch import NotFoundError
+from fastapi import APIRouter, HTTPException, status
+from fastapi.responses import JSONResponse
+
+from tds.db.elasticsearch import es_client
+from tds.modules.dataset.model import Dataset
+from tds.operation import create, delete, retrieve, update
+from tds.settings import settings
+
+dataset_router = APIRouter()
+logger = Logger(__name__)
+es_index = Dataset.index
+
+if settings.STORAGE_HOST:
+    s3 = boto3.client(
+        "s3",
+        endpoint_url=settings.STORAGE_HOST,
+        aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+        aws_session_token=None,
+        config=boto3.session.Config(signature_version="s3v4"),
+        verify=False,
+    )
+else:
+    s3 = boto3.client("s3")
+es = es_client()
+
+
+@dataset_router.get("", **retrieve.fastapi_endpoint_config)
+def list_datasets(page_size: int = 100, page: int = 0) -> List[Dataset]:
+    """
+    Retrieve the list of models from ES.
+    """
+    list_body = {
+        "size": page_size,
+    }
+    if page != 0:
+        list_body["from"] = page
+    res = es.search(index=es_index, **list_body)
+    return [Dataset(**hit["_source"]) for hit in res["hits"]["hits"]]
+
+
+@dataset_router.post("", **create.fastapi_endpoint_config)
+def create_dataset(payload: Dataset) -> JSONResponse:
+    """
+    Create model and return its ID
+    """
+    res = payload.create()
+    logger.info("New dataset created: %s", id)
+    return JSONResponse(
+        status_code=201,
+        content={"id": res["_id"]},
+    )
+
+
+@dataset_router.get("/{dataset_id}", **retrieve.fastapi_endpoint_config)
+def dataset_get(dataset_id: str | int) -> Dataset:
+    """
+    Returns a full representation of a Dataset
+    """
+    try:
+        res = es.get(index=es_index, id=dataset_id)
+        logger.info("dataset retrieved: %s", dataset_id)
+
+        return Dataset(**res["_source"])
+    except NotFoundError as exc:
+        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND) from exc
+
+
+@dataset_router.put("/{dataset_id}", **update.fastapi_endpoint_config)
+def dataset_put(dataset_id: str | int, payload: Dataset) -> JSONResponse:
+    """
+    Update dataset with full payload
+    """
+    if payload.id != dataset_id:
+        raise HTTPException(
+            status_code=422, detail="ID in request URL and in payload must match."
+        )
+    res = payload.save(dataset_id)
```

Comment:
This ID needs to be removed as the save no longer takes an argument.

Reply:
Thanks! Fixed.